File: cpp/ibcmer.cpp
// IBCMer - C++
//   URL: https://github.com/pennbauman/ibcmer
//   License: Creative Commons Attribution Share Alike 4.0 International
//   Author: Penn Bauman <me@pennbauman.com>
#define VERSION 0.3

#include <iostream>
#include <string>
#include <stack>

#include "src/program.h"
#include "src/doubleByte.h"
using namespace std;

// Check if string is a number
bool isNum(string s) {
	for (unsigned int i = 0; i < s.size(); i++) {
		if (!isdigit(s.at(i)))
			return false;
	}
	return true;
}
// Print help menu
void printHelp() {
	cout << "Usage: ibcmer [file] [options]" << endl;
	cout << "  --version        Print version and memory information." << endl;
	cout << "  --help, -h       Print this help menu." << endl;
	cout << "  --strict         Strictly follow format rules." << endl;
	cout << "  --check, -c      Check line numbers." << endl;
	cout << "  -b <numbers>     Sets break-points (listed in hex)." << endl;
	cout << "  --step           Starts program in debug mode." << endl;
	cout << "  --quiet          Don't print detailed output." << endl;
	cout << endl;
	cout << "Debug Environment: " << endl;
	cout << "  run                    Exit debugging and run program." << endl;
	cout << "  step [number]          Runs n commands then return to debug," << endl;
	cout << "                           defaults to running one command." << endl;
	cout << "  view [address|all]     Print contents of given memory address," << endl;
	cout << "                           all occupied memory if nothing is specified," << endl;
	cout << "                           or all memory 'all' is specified." << endl;
	cout << "  set <address> <value>  Set memory at given address to value." << endl;
	cout << "  exit                   Ends program immediately." << endl;
}

int main(int argc, char* argv[]) {
	// Check if short ints are 2 bytes
	if (sizeof(short int) != 2) {
		cerr << "PANIC: data values are the wrong size." << endl;
	}
	// Check parameters are given
	if (argc < 2) {
		cerr << "ERROR: Requires parameters." << endl;
		cerr << endl;
		printHelp();
		return 1;
	}
	// Check for stand alone parameters
	string current;
	for (int i = 1; i < argc; i++) {
		current = argv[i];
		if ((current == "--help") || (current == "-h")) {
			printHelp();
			return 0;
		}
		if (current == "--version") {
			cout << "IBCMer version " << fixed << setprecision(1) << VERSION << endl;
			return 0;
		}
	}
	// Interate through parameters
	bool strict = false;
	bool check = false;
	int step = -1;
	int loud = true;
	doubleByte temp;
	stack<int>* breaks = new stack<int>();
	for (int i = 2; i < argc; i++) {
		current = argv[i];
		if (current == "--strict") {
			strict = true;
		} else if ((current == "--check") || (current == "-c")) {
			strict = true;
			check = true;
		} else if (current == "-b") {
			if (!checkHex(argv[i+1])) {
				cerr << "ERROR: -b requires at least one break-point, break-points must be numbers." << endl;
				return 1;
			}
			do {
				i++;
				temp.setVal(argv[i]);
				breaks->push(temp.uint());
				if (i + 1 == argc)
					break;
			} while (checkHex(argv[i+1]));
		} else if (current == "--step") {
			step = 1;
		} else if ((current == "--quiet") || (current == "-q")) {
			loud = false;
		} else {
			cerr << "ERROR: Unknown parameter '" << argv[i] << "'" << endl;
			return 1;
		}
	}
	// Initialize program and memory
	string code = argv[1];
	program p = program();
	if (!p.init(code, breaks, strict, check)) {
		return 1;
	}
	// Run program
	int result = 0;
	while (true) {
		step--;
		result = p.step((loud || (step >= 0)));
		// Check for halt
		if (result == 0)
			break;
		// Check for breakpoint and go into debug mode
		if ((result == 2) || (step == 0)) {
			string input = "";
			while (true) {
				cout << "~ ";
				getline(cin, input);
				if (input == "run") {
					step = -1;
					break;
				} else if (input == "step") {
					step = 1;
					break;
				} else if (input.substr(0,4) == "step") {
					if (isNum(input.substr(5))) {
						step = atoi(input.substr(5).c_str());
						break;
					} else {
						cerr << " Step requires a number." << endl;
					}
				} else if (input == "view") {
					p.print(false);
				} else if (input == "view all") {
					p.print(true);
				} else if (input.substr(0,4) == "view") {
					if ((checkHex(input.substr(5))) && (input.size() < 9)) {
						temp = doubleByte(input.substr(5));
						p.printMem(temp.uint());
					}
				} else if (input.substr(0,3) == "set") {
					if (input.size() < 7) {
						cerr << " Set requires an address and a value." << endl;
					}
					for (unsigned int i = 1; i < input.size()-4; i++) {
						if (input.at(4+i) == ' ') {
							if ((i < 4) && (input.size()-5-i < 5) && checkHex(input.substr(4, i)) && checkHex(input.substr(5+i))) {
								doubleByte temp1 = doubleByte(input.substr(4, i));
								doubleByte temp2 = doubleByte(input.substr(5+i));
								if (!p.setMem(temp1.uint(), temp2.uint())) {
									cerr << " Invalid memory address." << endl;
								}
							} else {
								cerr << " Improperly formated address or value." << endl;
							}
							break;
						}
						if (i == input.size()-5)
							cerr << " Set requires an address and a value." << endl;
					}
				} else if (input == "exit") {
					return 0;
				} else {
					cerr << " Unknown command." << endl;
				}
			}
		}
		// Check for errors
		if (result == 3) {
			break;
		}
	}
	return 0;
}

File: R/ibcmer.R
#!/usr/bin/env Rscript
# IBCMer - R
#   URL: https://github.com/pennbauman/ibcmer
#   License: Creative Commons Attribution Share Alike 4.0 International
#   Author: Penn Bauman <me@pennbauman.com>
library("bitops")
MEM_SIZE <- 4096


MEMORY <- array(0, MEM_SIZE)
ACC <- 0
PC <- 1


# Read code file
input_file <- commandArgs(trailingOnly=TRUE)
i <- 1
for (l in readLines(input_file)) {
	MEMORY[i] <- strtoi(substr(l, 0, 4), 16)
	i <- i + 1
}
stdin <- file("stdin", "r")


# Execute code
while (TRUE) {
	opcode <- bitShiftR(MEMORY[PC], 12)
	address <- bitwAnd(MEMORY[PC], 0x0fff)
	cat(sprintf("[%03x]%04x  ", PC - 1, MEMORY[PC]))

	# halt
	if (opcode == 0x0) {
		cat("halt\n")
		break
	# i/o
	} else if (opcode == 0x1) {
		subopcode = bitShiftR(address, 8)
		cat(sprintf("i/o   (ACC)%04x\n", ACC))
		if (subopcode == 0x0) {
			cat("Input hex:  ")
			text <- readLines(stdin, n=1)
			if (nchar(text) > 4) {
				stop(sprintf("Invalid hex input '%s'", text))
			}
			ACC <- strtoi(text, 16)
		} else if (subopcode == 0x4) {
			cat("Input char: ")
			text <- readLines(stdin, n=1)
			if (nchar(text) != 1) {
				stop("A single character must be entered")
			}
			ACC <- utf8ToInt(text)
		} else if (subopcode == 0x8) {
			cat(sprintf("Output hex:  %04x\n", ACC))
		} else if (subopcode == 0xc) {
			cat(sprintf("Output char: %s\n", intToUtf8(ACC)))
		} else {
			write(sprintf("Unknown i/o sub-opcode '%x'", subopcode), stderr())
			quit(status = 1)
		}
	# shift
	} else if (opcode == 0x2) {
		subopcode = bitShiftR(address, 8)
		distance <- bitwAnd(address, 0xf)
		if (subopcode == 0x0) {
			res = bitwAnd(bitShiftL(ACC, distance), 0xffff)
			arrow <- "<<"
		} else if (subopcode == 0x4) {
			res = bitwAnd(bitShiftR(ACC, distance), 0xffff)
			arrow <- ">>"
		} else if (subopcode == 0x8) {
			res = bitwAnd(bitwOr(bitShiftL(ACC, distance), bitShiftR(ACC, (16 - distance))), 0xffff)
			arrow <- "<="
		} else if (subopcode == 0xc) {
			res = bitwAnd(bitwOr(bitShiftR(ACC, distance), bitShiftL(ACC, (16 - distance))), 0xffff)
			arrow <- "=>"
		} else {
			write(sprintf("Unknown shift sub-opcode '%x'", subopcode), stderr())
			quit(status = 1)
		}
		cat(sprintf("shift (ACC)%04x = (ACC)%04x %s %x\n", res, ACC, arrow, distance))
		ACC <- res
	# load
	} else if (opcode == 0x3) {
		ACC <- MEMORY[address + 1]
		cat(sprintf("load  (ACC)%04x = [%03x]%04x\n", ACC, address, ACC))
	# store
	} else if (opcode == 0x4) {
		MEMORY[address + 1] <- ACC
		cat(sprintf("store [%03x]%04x = (ACC)%04x\n", address, ACC, ACC))
	# add
	} else if (opcode == 0x5) {
		val <- MEMORY[address + 1]
		res <- bitwAnd(ACC + val, 0xffff)
		cat(sprintf("add   (ACC)%04x = (ACC)%04x + [%03x]%04x\n", res, ACC, address, val))
		ACC <- res
	# sub
	} else if (opcode == 0x6) {
		val <- MEMORY[address + 1]
		res <- ACC - val
		if (res < 0) {
			res <- res + 0x10000
		}
		cat(sprintf("sub   (ACC)%04x = (ACC)%04x - [%03x]%04x\n", res, ACC, address, val))
		ACC <- res
	# and
	} else if (opcode == 0x7) {
		val <- MEMORY[address + 1]
		res <- bitwAnd(ACC, val)
		cat(sprintf("and   (ACC)%04x = (ACC)%04x & [%03x]%04x\n", res, ACC, address, val))
		ACC <- res
	# or
	} else if (opcode == 0x8) {
		val <- MEMORY[address + 1]
		res <- bitwOr(ACC, val)
		cat(sprintf("or    (ACC)%04x = (ACC)%04x | [%03x]%04x\n", res, ACC, address, val))
		ACC <- res
	# xor
	} else if (opcode == 0x9) {
		val <- MEMORY[address + 1]
		res <- bitwXor(ACC, val)
		cat(sprintf("xor   (ACC)%04x = (ACC)%04x ^ [%03x]%04x\n", res, ACC, address, val))
		ACC <- res
	# not
	} else if (opcode == 0xa) {
		res = bitwAnd(bitwNot(ACC), 0xffff)
		cat(sprintf("not   (ACC)%04x = ! (ACC)%04x\n", res, ACC))
		ACC <- res
	# nop
	} else if (opcode == 0xb) {
		cat("nop\n")
	# jmp
	} else if (opcode == 0xc) {
		cat(sprintf("jmp   [%03x]\n", address))
		PC <- address
	# jmpe
	} else if (opcode == 0xd) {
		if (ACC == 0) {
			cat(sprintf("jmpe  [%03x]\n", address))
			PC <- address
		} else {
			cat(sprintf("jmpe  (ACC)%04x\n", ACC))
		}
	# jmpl
	} else if (opcode == 0xe) {
		if (bitShiftR(ACC, 15) == 1) {
			cat(sprintf("jmpl  [%03x]\n", address))
			PC <- address
		} else {
			cat(sprintf("jmpl  (ACC)%04x\n", ACC))
		}
	# brl
	} else if (opcode == 0xf) {
		ACC <- PC
		PC <- address
		cat(sprintf("brl   [%03x]  (ACC)%04x\n", address, ACC))
	} else {
		write("Unknown opcode", stderr())
		quit(status = 1)
	}

	PC <- PC + 1
	if (PC > MEM_SIZE) {
		write(sprintf("Error: Memory overflow (PC = 0x%04x)", PC - 1), stderr())
		quit(status = 1)
	}
}
